### DISCUSSION TITLE
Checkout options for fast-forward merge

### DISCUSSION THREAD

**Discussion by Uncommon** [libgit2/libgit2#6105] [Q&A]
I'm trying to implement a fast-forward merge, and I'm having trouble with the checkout part of it. I have two test cases where one fails if I use `GIT_CHECKOUT_SAFE` and the other fails if I use `GIT_CHECKOUT_FORCE`. In `perform_fastforward()` from examples/merge.c, it uses safe. Both tests pass if I merge using the command line `git merge otherBranch`.

The case that needs `GIT_CHECKOUT_FORCE` is where a file has been modified in the incoming branch. With safe checkout, the change is lost.

The case that needs `GIT_CHECKOUT_SAFE` is when a new file has been added to the index before merging. With forced checkout, the index change is lost.

Any advice?

**Comment by ethomson** [comment]
It's hard to speculate.  `SAFE` should be fine in the general case and emulate `git merge` on the command line.  Can you provide a simple repro case?

**Reply by ethomson** [reply]
Also, when you say that you have two test cases that fail, does that mean that two out of 50 fail, or you only have two and they both fail?

**Reply by Uncommon** [reply]
I have several merge tests, mostly based on Git's unit tests. The second test I mentioned, with a new staged file, is one that I just added because I discovered that scenario wasn't working in my app. Changing the checkout options made that test pass but one of the other tests failed.

My app is written in Swift with various wrapper functions, but my ff merge function should be mostly legible:
https://github.com/Uncommon/Xit/blob/master/Xit/Repository/XTRepository%2BMergePushPull.swift#L89

I'll see about coming up with a simple test case. Thanks for the quick responses.

**Reply by Uncommon** [reply]
Here is a rough draft of a test to explore this issue. I haven't actually run it yet; I need to add the test repo as well. It's been a long time since I last actually ran the libgit2 tests :)  Hopefully this can at least give you an idea of what's going on for me.

https://github.com/Uncommon/libgit2/commit/10e6b6aa16eece5ee6fc21537e3f6651b3ef7f05

**Reply by ethomson** [reply]
Thanks!  I'll take a look this weekend.

**Reply by Uncommon** [reply]
I realized after posting the link that there are a couple of obvious mistakes in the code. I did say it was a rough draft :)  I hope to have a chance to fix it up soon.
